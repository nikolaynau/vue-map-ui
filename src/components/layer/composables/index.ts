export * from './injectionSymbols';
export * from './useTileLayer';
export * from './useMarker';
export * from './useIcon';
export * from './useDivIcon';
