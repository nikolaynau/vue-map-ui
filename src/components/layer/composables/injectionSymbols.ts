import type { InjectionKey, Ref } from 'vue';
import type { DivIcon, Icon, Marker, TileLayer } from 'leaflet';
import type { UseMarkerApiReturn } from './useMarkerApi';

export const tileLayerKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'TileLayer'
    : ''
) as InjectionKey<Ref<TileLayer | null>>;

export const markerKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'Marker'
    : ''
) as InjectionKey<Ref<Marker | null>>;

export const iconKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'Icon'
    : ''
) as InjectionKey<Ref<Icon | null>>;

export const divIconKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'DivIcon'
    : ''
) as InjectionKey<Ref<DivIcon | null>>;

export const markerApiKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'MarkerApi'
    : ''
) as InjectionKey<UseMarkerApiReturn>;
