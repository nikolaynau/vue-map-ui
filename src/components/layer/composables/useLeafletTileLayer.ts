import type { MaybeReadonlyRef, MaybeRef } from '@vueuse/shared';
import type { Map, TileLayer, TileLayerOptions } from 'leaflet';
import { ref, type Ref } from 'vue';

export type UseTileLayerOptions = TileLayerOptions;

export function useLeafletTileLayer(
  map: MaybeRef<Map | null | undefined>,
  url: MaybeRef<string | null | undefined>,
  options: UseTileLayerOptions = {}
): Ref<TileLayer | null> {
  const tileLayer = ref<TileLayer | null>(null) as Ref<TileLayer | null>;
  return tileLayer;
}

export type UseLeafletTileLayerReturn = ReturnType<typeof useLeafletTileLayer>;
