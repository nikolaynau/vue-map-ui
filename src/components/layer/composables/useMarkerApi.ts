import { type Ref } from 'vue';
import type { DivIcon, Icon } from 'leaflet';

export interface UseMarkerApiOptions {
  icon: Ref<Icon | DivIcon | null | undefined>;
}

export function useMarkerApi(options: UseMarkerApiOptions) {
  function setIcon(icon: Icon | DivIcon | null | undefined) {
    options.icon.value = icon;
  }

  return {
    setIcon
  };
}

export type UseMarkerApiReturn = ReturnType<typeof useMarkerApi>;
