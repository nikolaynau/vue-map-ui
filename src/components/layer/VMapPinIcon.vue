<script setup lang="ts">
import {
  computed,
  reactive,
  ref,
  toRefs,
  useSlots,
  type StyleValue
} from 'vue';
import type { DivIcon } from 'leaflet';
import { useTemplateRef } from '../../composables';
import {
  default as VMapDivIcon,
  type Attrs as IconAttrs,
  type Props as IconProps
} from './VMapDivIcon.vue';

export interface Props extends IconProps {
  color?: string;
  backgroundColor?: string;
  placeholderColor?: string;
}

export type Attrs = IconAttrs;

const props = withDefaults(defineProps<Props>(), {
  iconSize: () => [32, 46],
  iconAnchor: () => [16, 46],
  color: undefined,
  backgroundColor: undefined,
  placeholderColor: undefined
});

const { color, backgroundColor, placeholderColor, ..._iconProps } =
  toRefs(props);
const iconProps = reactive(_iconProps);
const slots = useSlots() as { default: unknown };
const popupAnchor = ref([0, -50]);

const pinStyle = computed<StyleValue>(() => ({
  color: backgroundColor.value ?? color.value
}));

const iconStyle = computed<StyleValue>(() => ({
  color: color.value,
  backgroundColor: placeholderColor.value
}));

const { templateRef, value: icon } = useTemplateRef<
  InstanceType<typeof VMapDivIcon>,
  DivIcon | null
>(obj => obj.icon);

defineExpose({
  icon
});
</script>

<template>
  <VMapDivIcon
    ref="templateRef"
    class="v-map-pin-icon"
    v-bind="iconProps"
    :popup-anchor="popupAnchor"
  >
    <div class="v-map-pin-icon__shadow"></div>
    <div class="v-map-pin-icon__pin" :style="pinStyle"></div>
    <div class="v-map-pin-icon__icon" v-if="slots.default" :style="iconStyle">
      <slot></slot>
    </div>
  </VMapDivIcon>
</template>

<style>
@import 'pin-icon';
</style>
