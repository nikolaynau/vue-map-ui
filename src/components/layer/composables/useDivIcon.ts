import { inject, provide, readonly, type Ref } from 'vue';
import type { DivIcon } from 'leaflet';
import { divIconKey } from './injectionSymbols';

export function provideDivIcon(divIcon: Ref<DivIcon | null>) {
  provide(divIconKey, readonly(divIcon) as Ref<DivIcon | null>);
}

export function useDivIcon(): Ref<DivIcon | null> | undefined {
  return inject(divIconKey, undefined);
}
