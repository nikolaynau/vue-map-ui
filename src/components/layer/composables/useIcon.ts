import { inject, provide, readonly, type Ref } from 'vue';
import type { Icon } from 'leaflet';
import { iconKey } from './injectionSymbols';

export function provideIcon(icon: Ref<Icon | null>) {
  provide(iconKey, readonly(icon) as Ref<Icon | null>);
}

export function useIcon(): Ref<Icon | null> | undefined {
  return inject(iconKey, undefined);
}
