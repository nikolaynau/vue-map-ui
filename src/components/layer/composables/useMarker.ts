import { inject, provide, readonly, type Ref } from 'vue';
import type { Marker } from 'leaflet';
import { markerKey } from './injectionSymbols';

export function provideMarker(marker: Ref<Marker | null>) {
  provide(markerKey, readonly(marker) as Ref<Marker | null>);
}

export function useMarker(): Ref<Marker | null> | undefined {
  return inject(markerKey, undefined);
}
