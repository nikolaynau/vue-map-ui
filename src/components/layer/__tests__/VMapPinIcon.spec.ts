import { describe, it, expect, vi } from 'vitest';
import { ref, reactive, h, defineComponent, onMounted, nextTick } from 'vue';
import { DivIcon } from 'leaflet';
import { mount } from '../../../../.test';
import { provideApi } from '../../../composables';
import { markerApiKey, useDivIcon } from '../composables';
import VMapPinIcon, { type Props } from '../VMapPinIcon.vue';

describe('VMapPinIcon', () => {
  it('should be expose instance', () => {
    expect.assertions(1);

    const Component = defineComponent({
      setup() {
        const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);

        onMounted(() => {
          expect(icon.value?.icon).toBeInstanceOf(DivIcon);
        });

        return () => h(VMapPinIcon, { ref: icon });
      }
    });

    mount(Component);
  });

  it('should be provide icon instance', () => {
    expect.assertions(1);

    const ChildComponent = defineComponent({
      setup() {
        const icon = useDivIcon();
        expect(icon?.value).toBeInstanceOf(DivIcon);
        return () => null;
      }
    });

    const Component = defineComponent({
      setup() {
        return () => h(VMapPinIcon, () => h(ChildComponent));
      }
    });

    mount(Component);
  });

  it('should be use marker api for set icon', () => {
    const markerApi = {
      setIcon: vi.fn()
    };

    const RootComponent = defineComponent({
      setup(props, { slots }) {
        provideApi(markerApiKey, markerApi);
        return () => h('div', undefined, slots);
      }
    });

    const Component = defineComponent({
      setup() {
        return () => h(RootComponent, () => h(VMapPinIcon));
      }
    });

    const vm = mount(Component);
    vm.unmount();

    expect(markerApi.setIcon.mock.calls).toHaveLength(2);
    expect(markerApi.setIcon.mock.calls[0][0]).toBeInstanceOf(DivIcon);
    expect(markerApi.setIcon.mock.calls[1][0]).toBeNull();
  });

  it('should be render html', async () => {
    const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);

    const Component = defineComponent({
      setup() {
        return () => h(VMapPinIcon, { ref: icon });
      }
    });

    mount(Component);

    await nextTick();

    expect(icon.value?.icon?.options.html).toMatchSnapshot();
  });

  it('should be render html with default slot', async () => {
    const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapPinIcon, { ref: icon }, () => h('i', { class: 'icon' }));
      }
    });

    mount(Component);

    await nextTick();

    expect(icon.value?.icon?.options.html).toMatchSnapshot();
  });

  it('should work with color', async () => {
    const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);
    const props = reactive<Props>({
      color: '#fff'
    });

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapPinIcon, { ref: icon, ...props }, () =>
            h('i', { class: 'icon' })
          );
      }
    });

    mount(Component);

    await nextTick();

    expect(icon.value?.icon?.options.html).toMatchSnapshot();
  });

  it('should work with background color', async () => {
    const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);
    const props = reactive<Props>({
      color: '#fff',
      backgroundColor: '#000'
    });

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapPinIcon, { ref: icon, ...props }, () =>
            h('i', { class: 'icon' })
          );
      }
    });

    mount(Component);

    await nextTick();

    expect(icon.value?.icon?.options.html).toMatchSnapshot();
  });

  it('should work with placeholder color', async () => {
    const icon = ref<InstanceType<typeof VMapPinIcon> | null>(null);
    const props = reactive<Props>({
      placeholderColor: 'green'
    });

    const Component = defineComponent({
      setup() {
        return () =>
          h(VMapPinIcon, { ref: icon, ...props }, () =>
            h('i', { class: 'icon' })
          );
      }
    });

    mount(Component);

    await nextTick();

    expect(icon.value?.icon?.options.html).toMatchSnapshot();
  });
});
