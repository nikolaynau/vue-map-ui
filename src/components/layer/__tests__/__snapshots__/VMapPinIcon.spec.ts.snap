// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VMapPinIcon > should be render html 1`] = `"<div class=\\"v-map-pin-icon__shadow\\"></div><div class=\\"v-map-pin-icon__pin\\"></div><!--v-if-->"`;

exports[`VMapPinIcon > should be render html with default slot 1`] = `"<div class=\\"v-map-pin-icon__shadow\\"></div><div class=\\"v-map-pin-icon__pin\\"></div><div class=\\"v-map-pin-icon__icon\\"><i class=\\"icon\\"></i></div>"`;

exports[`VMapPinIcon > should work with background color 1`] = `"<div class=\\"v-map-pin-icon__shadow\\"></div><div class=\\"v-map-pin-icon__pin\\" style=\\"color: rgb(0, 0, 0);\\"></div><div class=\\"v-map-pin-icon__icon\\" style=\\"color: rgb(255, 255, 255);\\"><i class=\\"icon\\"></i></div>"`;

exports[`VMapPinIcon > should work with color 1`] = `"<div class=\\"v-map-pin-icon__shadow\\"></div><div class=\\"v-map-pin-icon__pin\\" style=\\"color: rgb(255, 255, 255);\\"></div><div class=\\"v-map-pin-icon__icon\\" style=\\"color: rgb(255, 255, 255);\\"><i class=\\"icon\\"></i></div>"`;

exports[`VMapPinIcon > should work with placeholder color 1`] = `"<div class=\\"v-map-pin-icon__shadow\\"></div><div class=\\"v-map-pin-icon__pin\\"></div><div class=\\"v-map-pin-icon__icon\\" style=\\"background-color: green;\\"><i class=\\"icon\\"></i></div>"`;
